google-cloud-firestore/src/main/java/com/google/cloud/firestore/RecursiveDelete.java

```diff
@@ -0,0 +1,353 @@
+            // If there are fewer than the number of documents specified in the limit() field, we
+            // know that the query is complete.
+            if (streamedDocsCount[0] < maxPendingOps) {
+              onQueryEnd();
+            } else if (pendingOperationsCount == 0) {
+              streamDescendants();
+            }
+          }
+        };
+
+    query.stream(responseObserver);
+  }
+
+  private Query getAllDescendantsQuery(@Nonnull CollectionReference collectionReference) {
+    // The parent is the closest ancestor document to the location we're deleting. Since we are
+    // deleting a collection, the parent is the path of the document containing that collection (or
+    // the database root, if it is a root collection).
+    ResourcePath parentPath = collectionReference.getResourcePath();
+    parentPath = parentPath.popLast();
+    String collectionId = collectionReference.getId();
+
+    Query query = baseAllDescendantsQuery(parentPath, collectionId);
+
+    // To find all descendants of a collection reference, we need to use a
+    // composite filter that captures all documents that start with the
+    // collection prefix. The MIN_ID constant represents the minimum key in
+    // this collection, and a null byte + the MIN_ID represents the minimum
+    // key is the next possible collection.
+    char nullChar = '\0';
+    String startAt = collectionId + "/" + REFERENCE_NAME_MIN_ID;
+    String endAt = collectionId + nullChar + "/" + REFERENCE_NAME_MIN_ID;
+    query =
+        query
+            .whereGreaterThanOrEqualTo(FieldPath.documentId(), startAt)
+            .whereLessThan(FieldPath.documentId(), endAt);
+
+    // startAfter() needs to be added after the where() filters since it creates an implicit
+    // orderBy.
+    if (lastDocumentSnap != null) {
+      query = query.startAfter(lastDocumentSnap);
+    }
+
+    return query;
+  }
+
+  private Query getAllDescendantsQuery(@Nonnull DocumentReference documentReference) {
+    // The parent is the closest ancestor document to the location we're deleting. Since we are
+    // deleting a document, the parent is the path of that document.
+    ResourcePath parentPath = documentReference.getResourcePath();
+    String collectionId = documentReference.getParent().getId();
+
+    return baseAllDescendantsQuery(parentPath, collectionId);
+  }
+
+  private Query baseAllDescendantsQuery(ResourcePath resourcePath, String collectionId) {
+    Query query =
+        new Query(
+            firestoreRpcContext,
+            QueryOptions.builder()
+                .setParentPath(resourcePath)
+                .setCollectionId(collectionId)
+                .setAllDescendants(true)
+                .setKindless(true)
+                .setRequireConsistency(false)
+                .build());
+
+    // Query for names only to fetch empty snapshots.
+    query = query.select(FieldPath.documentId()).limit(maxPendingOps);
+
+    return query;
+  }
```

Comment:
We should find a way to combine all three methods (by passing ResourcePath and collectionId at each callsite)